### ISSUE TITLE
Build script first steps: getting a list of source files

### ISSUE THREAD

**Issue by awvwgk** [fortran-lang/fpm#457]
To get started with the (two-way) build script, let's begin with something simple to explore how we can implement the basic required infrastructure.

Getting a list of source files, that is something fpm usually does automatically, but maybe if you have submodules which implement different functionality like in the [Focal](https://github.com/lkedward/focal) project this could be needed.

The output of the build script could look like this (maybe we find better/shorter keywords):

```
fpm:library-source="src/lib.f90"
fpm:app-<name>-source="app/util.f90"
fpm:app-<name>-main="app/main.f90"
fpm:demo-<name>-source="example/util.f90"
fpm:demo-<name>-main="example/main.f90"
fpm:test-<name>-source="test/util.f90"
fpm:test-<name>-main="test/main.f90"
...
```

The input would be given by environment variables or maybe command line arguments.

```
FPM_COMPILER=gfortran
FPM_BUILD_PROFILE=debug
...
```

In case of Focal the build script could decide depending on the value of `FPM_BUILD_PROFILE` which submodule source files to compile.

For the initial implementation we could just rely on the build script being executable, which would already allow shell, python, ruby, ... scripts as input. The input for the script would be in the build table:

```toml
[build]
script = "build.py"
```

**Comment by LKedward** [comment]
This is a good place to start and your proposed methodology looks good to me. I'm not too bothered about the verbosity of the keywords because presumably they can be abstracted away in functions.

---

With regards to *Focal*, in the short term I will likely replace the mutually exclusive submodules with a single submodule containing `#ifdef DEBUG` directives just to get it working with *fpm*. In the long term I had wondered whether *fpm* might eventually support selective inclusion/exclusion of sources based on selected profile/features/OS _etc._. I think this would be beneficial because Fortran submodules are well suited to this kind of plug-and-play of implementation.

**Comment by awvwgk** [comment]
> I had wondered whether _fpm_ might eventually support selective inclusion/exclusion of sources based on selected profile/features/OS _etc._

Submodules seem like the right mechanism for conditional compilation without having to rely on preprocessor, still I wonder how an automatic selection rule would look like for this purpose. Are you aware of any existing convention for submodule naming regarding this?

**Comment by ivan-pi** [comment]
While not exactly conditional compilation, I also found submodules very effective to switch between different algorithms/implementations while benchmarking the ascii character classification routines (see https://github.com/ivan-pi/fortran-ascii/blob/09c26de1cf00119eb5e8c3316e8163fa1677b3ca/Makefile#L39). Thinking of `stdlib`, I envisioned that in the future it would be possible to switch between different implementations/backends using submodules.

**Comment by LKedward** [comment]
> Submodules seem like the right mechanism for conditional compilation without having to rely on preprocessor, still I wonder how an automatic selection rule would look like for this purpose. Are you aware of any existing convention for submodule naming regarding this?

Yes I agree submodules are preferable for this than using a preprocessor, but I am not aware of any existing naming conventions or selection methods for doing this. Should we open a new issue to discuss this kind of conditional compilation?

**Comment by ivan-pi** [comment]
Just found another use case for submodules (I can provide a pdf copy if needed):
> Bader, R. (2006, August). Dynamic polymorphism in Fortran 95: how submodules can help. In ACM SIGPLAN Fortran Forum (Vol. 25, No. 2, pp. 3-7). https://doi.org/10.1145/1148105.1148106

**Comment by arjenmarkus** [comment]
Please do provide that. I have a paper copy somewhere, but a PDF version
provides easier access :).

Op ma 26 apr. 2021 om 12:42 schreef Ivan Pribec ***@***.***>:

**Comment by urbanjost** [comment]
Another option I miss, but only in a few projects, is conditional source directory selection. That is, when you have abstracted very system-specific routines with your own procedures so the file and function names are the same but there is very little if any duplicated code it is nice to say to build src/Microsoft instead of src/Solaris or src/BSD or src/Linux, where each directory contains the system-specific source for a particular platform.  Just something else to add to the wish-list :>
